### PR TITLE
feat(Shape): add Shape component

### LINKED ISSUE #196: <Shape /> Component
http://semantic-ui.com/modules/shape.html

### PR DESCRIPTION
**Pull request by tarang9211**
Fixes #196

### CODE REVIEW

**Comment by tarang9211** [comment]
@layershifter Could you guide me on this one, if possible? Thanks!

**Comment by tarang9211** [comment]
@layershifter Could you give me some feedback when you get a chance?

File: src/modules/Shape/Shape.js [hunk 1]
```diff
@@ -0,0 +1,51 @@
+import cx from 'classnames'
+import React, { PropTypes } from 'react'
+
+import {
+  AutoControlledComponent as Component,
```
Comment (layershifter) [inline src/modules/Shape/Shape.js]: Let's don't use `AutoControlledComponent` at now, let's back to `React.Component`.

**Review by layershifter — CHANGES_REQUESTED** [review]

**Comment by layershifter** [comment]
Let's start from `Shape` sub-components and some props:

```jsx
<Shape.Side active header>Did you know? This side starts visible.</Shape.Side>
//<div class="active ui header side">Did you know? This side starts visible.</div>
```

```jsx
<Shape.Side>
  <Shape.Content center>4</Shape.Content>
</Shape.Side>

<Shape.Side>
  <Shape.Content center='4' />
</Shape.Side>

<Shape.Side center content='4' />

//<div class="side">
//  <div class="content">
//    <div class="center">4</div>
//  </div>
//</div>
```

```jsx
<Shape cube />
//<div class="ui cube shape"></div>
```

```jsx
<Shape type='auto' />
//<div class="ui auto shape"></div>
<Shape type='square' />
//<div class="ui square shape"></div>
<Shape type='auto' />
//<div class="ui irregular shape"></div>
```

**Comment by tarang9211** [comment]
@layershifter Can we come up with a list of sub-components and props? I think that would make this easier!

**Comment by ch3lo** [comment]
I suggest you an update in readme.md marking shape as supported.

**Comment by levithomason** [comment]
@tarang9211 please see https://github.com/Semantic-Org/Semantic-UI-React/pull/1323#issuecomment-281065710 for subcomponents and props.

**Comment by tarang9211** [comment]
@levithomason yup, will have a look. I'm just swamped in the week, will be closing out my open issues by Saturday.

**Comment by cemremengu** [comment]
Any updates on this ? :eyes:

**Comment by madwill** [comment]
Hello everyone! Would love to see this component make it into the react implementation! Thank you all for your hard work ! Making the world a prettier place!

**Comment by levithomason** [comment]
Note, I've just rebased this branch to the latest master to keep it from getting to stale.

**Comment by madwill** [comment]
Has this been forgotten ?

**Comment by levithomason** [comment]
It appears @tarang9211 has gotten busy.  We'll gladly take another PR picking up where this one left off.

**Comment by madwill** [comment]
Thank you for such a quick response, I pulled the fork from tarang but unless i'm not understanding something i don't think we were that close to completion. 

For better understanding of the library and its component, i'll try and give it a shot. Really just at initial stage but i got it to show in the docs with example having shape components and 2 sub components being side and content. 

https://github.com/madwill/Semantic-UI-React/commit/12dc99dd09136a64102d82ac7527bda9dc9ffa25

I confess having currently no idea how to manipulate it to get the semantic ui shape behavior. I'll dig into the official code for it .shape() and .shape('flip X'); as well as other components using javascript for behavior within Semantic-UI-React to see how it already handle such cases. 

I can not make promises as i don't feel confident enough right now but if you could point to a component you feel would be a good example of javascript manipulation without jquery it it may accelerate / make it possible. 

Thank you very much

**Comment by tarang9211** [comment]
Hey guys, I'm really sorry for being away! Been extremely busy with work and personal commitments. Happy to help whenever needed.

**Comment by madwill** [comment]
Turns out is a surprisingly complex component :

https://github.com/Semantic-Org/Semantic-UI/blob/master/src/definitions/modules/shape.js

The way it invokes methods, using query arguments in order to allow for a .shape('flip up') makes it so that its not just re-implementation, it would require some re-engineering. 

With the 6 types of animations requiring adjustments based on component size, its a longer task than expected. 

I get why it stalled. Jquery less semantic UI is an ambitious project considering its implementation. It would even say it has to lose a bit of its "semantic-ism". 

This feels like a lot for a newcomer so i'll let my head cool down.   I don't think i'll be the one creating the pull request. 

Perhaps the flip / stage section will become a bit more clear as i take a step back. 

Thanks for such big amount of work.

**Comment by madwill** [comment]
What if we simplify it a bit ? for the sake of quick implement-ability.

Top component (shape) would have 2 props being side and direction

```

    /** Currently displayed side, */
    side: PropTypes.number,

     /** Direction the shape should rotate. */
    direction: PropTypes.oneOf(['right', 'left', 'up', 'down', 'over', 'back']),
```
Usage would be something like this 

```
       <Shape type='auto' side={1} direction='left'>
          <Shape.Side >
            <Image src='https://react.semantic-ui.com/assets/images/avatar/large/matthew.png' />
          </Shape.Side>
          <Shape.Side>
            <Image src='https://react.semantic-ui.com/assets/images/avatar/large/elliot.jpg' />
          </Shape.Side>
        </Shape>
```

User would have to change side by himself with his own controls and state.

We would trigger animation in a life cycle function such as 

```
 componentWillReceiveProps(nextProps) {
    if (nextProps.side !== this.props.side) {
      this.startAnimating()
    }
  }

```
User would still have the liberty to change the "direction" props if that's something someone would want and i would have to make a big "switch" statement in startAnimating to create the correct animation depending on our props and type 

What would be the point of 

`<Shape.Content center='4' />`

It feels like an extra layer that could be avoided, if you look at usage within semantic-ui itself there is no side-content divs. Content is directly written in the side div. 

```
<div class="ui shape">
  <div class="sides">
    <div class="active side">This side starts visible.</div>
    <div class="side">This is yet another side</div>
    <div class="side">This is the last side</div>
  </div>
</div>

```
Let me know what you think

**Comment by layershifter** [comment]
@madwill Thanks for a detailed suggestions :+1:

My initial idea was about `componentWillReceiveProps`, too. I'll be glad to see changes that realizing this idea.

**Comment by layershifter** [comment]
I take the PR on myself, #1751

**Pull request closed without merging**